sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala

```diff
@@ -19,39 +19,193 @@ package org.apache.spark.sql.catalyst
 
 import org.apache.spark.sql.Row
 import org.apache.spark.sql.catalyst.expressions.GenericRow
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.unsafe.types.UTF8String
 
 /**
  * An abstract class for row used internal in Spark SQL, which only contain the columns as
  * internal types.
+ *
+ * The following is a mapping between Spark SQL types and types of objects in row:
+ *
+ *   BooleanType -> java.lang.Boolean
+ *   ByteType -> java.lang.Byte
+ *   ShortType -> java.lang.Short
+ *   IntegerType -> java.lang.Integer
+ *   FloatType -> java.lang.Float
+ *   DoubleType -> java.lang.Double
+ *   StringType -> UTF8String
+ *   DecimalType -> org.apache.spark.sql.types.Decimal
+ *
+ *   DateType -> java.lang.Int
+ *   TimestampType -> java.lang.Long
+ *
+ *   BinaryType -> Array[Byte]
+ *   ArrayType -> scala.collection.Seq
+ *   MapType -> scala.collection.Map
+ *   StructType -> InternalRow
  */
-abstract class InternalRow extends Row {
-  // A default implementation to change the return type
-  override def copy(): InternalRow = {this}
-}
+abstract class InternalRow extends Serializable {
+  /** Number of elements in the Row. */
+  def size: Int = length
```

Comment:
agree. It's internal and we don't need to follow scala convention to provide `size`.